File: 02_Patterns/Pattern_20/Pattern_20.py
class Solution:
    def printTriangle(self, N):
        # outer loop for number of rows/lines
        for i in range(1, 2*N):

            # stars would be equal to the row no. until first half
            stars = i

            # initialise spaces
            spaces = 2*N-2*i

            # for the second half of the rotated triangle.
            if i > N:
                stars = 2*N-i
                spaces = 2*i-2*N

            # stars
            for j in range(1, stars+1):
                print("*", end="")

              # spaces
            for j in range(1, spaces+1):
                print(" ", end="")

              # stars
            for j in range(1, stars+1):
                print("*", end="")
            print()


# {
# Driver Code Starts
# Initial Template for Python 3
if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        N = int(input())
        ob = Solution()
        ob.printTriangle(N)
# } Driver Code Ends

File: 02_Patterns/Pattern_20/Pattern_20.js
function butterflyPattern(n) {
  // Outer loop for number of rows/lines
  for (let i = 1; i < 2 * n; i++) {
    let stars, spaces;

    // Calculate stars and spaces based on the row number
    // For upper half of the pattern
    if (i <= n) {
      stars = i;
      spaces = 2 * n - 2 * i;
    }
    // for lower half of the pattern
    else {
      stars = 2 * n - i;
      spaces = 2 * i - 2 * n;
    }

    let row = '';

    // Print stars
    for (let j = 1; j <= stars; j++) {
      row += '*';
    }

    // Print spaces
    for (let j = 1; j <= spaces; j++) {
      row += ' ';
    }

    // Print stars again
    for (let j = 1; j <= stars; j++) {
      row += '*';
    }

    // Output the row
    console.log(row);
  }
}

// call the function
// butterflyPattern(5);

// Export the function for testing
module.exports = butterflyPattern;
